### PR TITLE
CDRIVER-3934 Correctly handle network errors during speculative authentication

### PR DESCRIPTION
**Pull request by alcaeus**
CDRIVER-3934

Also adds an assertion to avoid memory access failure that appeared once with VS15

### CODE REVIEW

File: src/libmongoc/tests/test-mongoc-client.c [hunk 1]
```diff
@@ -484,6 +484,138 @@ test_mongoc_client_authenticate (void *context)
...
+   /*
+    * Log in as admin.
+    */
+   admin_client = test_framework_client_new ();
+
+   /*
+    * Add a user to the test database.
+    */
+   username = gen_test_user ();
+   uri = gen_good_uri (username, "test");
```
Comment (kevinAlbs) [inline src/libmongoc/tests/test-mongoc-client.c:516]: Unused

File: src/libmongoc/tests/test-mongoc-client.c [hunk 2]
```diff
@@ -484,6 +484,138 @@ test_mongoc_client_authenticate (void *context)
...
+                "'mode': {'times': 1}, "
+                "'data': {'failCommands': ['saslContinue'], 'closeConnection': "
+                "true, 'errorCode': 10107}}"),
+      NULL,
+      NULL,
+      &error);
+   ASSERT_OR_PRINT (r, error);
+   mongoc_database_destroy (database);
+
+   /* Try authenticating by creating a user */
```
Comment (kevinAlbs) [inline src/libmongoc/tests/test-mongoc-client.c:567]:
```suggestion
   /* Try authenticating by creating a user */
```
```suggestion
   /* Try authenticating by running a find operation */
```

**Review by kevinAlbs — APPROVED** [review]
LGTM!

**Comment by alcaeus** [comment]
Build failures are unrelated:
* asan-ubuntu is failing on master
* versioned-api is failing on master
* windows-2015 is a ping failure to an atlas cluster (I suspect that one of our evergreen project variables points to the wrong thing).

**Comment by alcaeus** [comment]
Cherry-picked into 1.17 in e657c13a4.

**Pull request merged**